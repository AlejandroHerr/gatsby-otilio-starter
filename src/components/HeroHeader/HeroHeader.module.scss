@import '../../colors';
@import '../../fonts';

.hero_header {
  position: relative;
  padding-top: 6em;
}

.hero_header__inner {
  max-width: 32em;
  margin: 0 auto;
  padding: 0 2em;
}

.hero_header--has_image {
  padding: 16em 0 2em;
  background: $color-black;
  text-shadow: 0 1px 2px rgba(0, 0, 0, .1);
  overflow: hidden;
}

.hero_image {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: no-repeat scroll center center / cover $color-black;
  animation: .5s ease-in forwards fadein;
  z-index: 5;

  img {
    margin: 0;
  }
}

.hero_image__shadow {
  position: absolute;
  bottom: 0;
  left: 0;
  width: 100%;
  height: 12em;
  background:
    linear-gradient(
      to bottom,
      rgba(0, 0, 0, 0) 0%,
      rgba(0, 0, 0, .5) 100%
    );
  z-index: 8;
}

@media only screen and (max-width: 480px) {
  .hero_header__inner {
    padding: 0 16px;
  }
}
