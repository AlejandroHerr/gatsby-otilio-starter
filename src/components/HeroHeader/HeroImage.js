// @flow
/* globals document */
import React, { PureComponent } from 'react';
import Img from 'gatsby-image';

import type { SizeImageSharpType } from '../../types/gatsby';

import styles from './HeroHeader.module.scss';

type PropsType = {
  image: SizeImageSharpType,
};

type StateType = {
  translateY: number,
};

export default class HeroImage extends PureComponent<PropsType, StateType> {
  handleScroll: () => void;

  constructor(props: PropsType) {
    super(props);

    this.state = {
      translateY: 0,
    };

    this.handleScroll = this.handleScroll.bind(this);
  }

  componentDidMount() {
    document.addEventListener('scroll', this.handleScroll);
  }

  componentWillUnmount() {
    document.removeEventListener('scroll', this.handleScroll);
  }

  handleScroll() {
    const scrollTop = (document.documentElement && document.documentElement.scrollTop)
      || 0;

    this.setState({ translateY: scrollTop / 5 });
  }
  render() {
    const { image } = this.props;
    const { translateY } = this.state;

    return (
      <div
        className={styles.hero_image}
        style={{ transform: `translate3d(0px, ${translateY}px, 0px)` }}
      >
        <Img
          sizes={image}
          style={{
            position: 'absolute',
            left: 0,
            top: 0,
            width: '100%',
            height: '100%',
            margin: '0',
          }}
        />
        <div className={styles.hero_image__shadow} />
      </div>
    );
  }
}
