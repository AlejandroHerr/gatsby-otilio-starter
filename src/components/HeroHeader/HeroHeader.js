// @flow
import React, { type Node } from 'react';
import classnames from 'classnames';

import HeroImage from './HeroImage';
import type { SizeImageSharpType } from '../../types/gatsby';

import styles from './HeroHeader.module.scss';

type PropsType = {
  children: Node,
  className?: string,
  image?: ?SizeImageSharpType,
};


const HeroHeader = ({ children, className, image }: PropsType) => (
  <header className={classnames(styles.hero_header, className, { [styles['hero_header--has_image']]: !!image })}>
    <div className={styles.hero_header__inner}>
      {children}
      {image && (
        <HeroImage
          image={image}
        />
      )}
    </div>
  </header>
);

HeroHeader.defaultProps = {
  className: '',
  image: null,
};

export default HeroHeader;
