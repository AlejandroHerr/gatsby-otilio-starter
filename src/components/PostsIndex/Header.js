// @flow
import React from 'react';
import classnames from 'classnames';

import type { SizeImageSharpType } from '../../types/gatsby';

import HeroHeader from '../HeroHeader';

import styles from './PostsIndex.module.scss';

type PropsType = {
  description: string,
  image: SizeImageSharpType,
  title: string,
};

const MainHeader = ({
  description, image, title,
}: PropsType) => (
  <HeroHeader
    className={classnames(styles.posts_index_header, {
        [styles['posts_index_header--has_image']]: !!image,
      })}
    image={image}
  >
    <h1
      className={classnames(styles.posts_index_header__title, {
          [styles['posts_index_header__title--has_image']]: !!image,
        })}
    >
      {title}
    </h1>
    <span
      className={classnames(styles.posts_index_header__body, {
          [styles['posts_index_header__body--has_image']]: !!image,
        })}
    >
      {description}
    </span>
  </HeroHeader>
);

export default MainHeader;
