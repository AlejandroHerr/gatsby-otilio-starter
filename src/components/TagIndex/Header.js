// @flow
import React from 'react';
import classnames from 'classnames';

import type { SizeImageSharpType } from '../../types/gatsby';

import HeroHeader from '../HeroHeader';

import styles from './TagIndex.module.scss';

type PropsType = {
  image?: ?SizeImageSharpType,
  title: string,
  total: number,
};

const MainHeader = ({
  image, title, total,
}: PropsType) => (
  <HeroHeader
    className={classnames(styles.tag_index_header, {
      [styles['tag_index_header--has_image']]: !!image,
    })}
    image={image}
  >
    <h1
      className={classnames(styles.tag_index_header__title, {
          [styles['tag_index_header__title--has_image']]: !!image,
        })}
    >
      {title}
    </h1>
    <span
      className={classnames(styles.tag_index_header__body, {
          [styles['tag_index_header__body--has_image']]: !!image,
        })}
    >
      {`${total} posts in this category`}
    </span>
  </HeroHeader>
);


MainHeader.defaultProps = {
  image: null,
};

export default MainHeader;
