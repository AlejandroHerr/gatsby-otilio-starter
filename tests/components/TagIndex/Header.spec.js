import React from 'react';
import { shallow } from 'enzyme';

import Header from '../../../src/components/TagIndex/Header';
import HeroHeader from '../../../src/components/HeroHeader';

const defaultProps = {
  total: 50,
  title: 'Test title',
};

const setup = (props = {}) => {
  const header = shallow(<Header {...defaultProps} {...props} />);

  return {
    header,
    heroHeader: header.find(HeroHeader),
  };
};

describe('TagIndex', () => {
  describe('Header', () => {
    it('should render a HeroHeader', () => {
      const { header, heroHeader } = setup();

      expect(header.find('h1')).toHaveLength(1);
      expect(header.find('h1').text()).toBe(defaultProps.title);
      expect(header.find('span')).toHaveLength(1);
      expect(header.find('span').text()).toBe(`${defaultProps.total} posts in this category`);
      expect(heroHeader).toHaveLength(1);
      expect(header).toMatchSnapshot();
    });
    it('should render a HeroHeader', () => {
      const { header, heroHeader } = setup({
        image: {
          size: {
            src: 'testSrc',
          },
        },
      });

      expect(heroHeader).toHaveLength(1);
      expect(header).toMatchSnapshot();
    });
  });
});
