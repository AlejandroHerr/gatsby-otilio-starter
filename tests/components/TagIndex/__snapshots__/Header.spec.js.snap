// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagIndex Header should render a HeroHeader 1`] = `
<HeroHeader
  className="tag_index_header"
  image={null}
>
  <h1
    className="tag_index_header__title"
  >
    Test title
  </h1>
  <span
    className="tag_index_header__body"
  >
    50 posts in this category
  </span>
</HeroHeader>
`;

exports[`TagIndex Header should render a HeroHeader 2`] = `
<HeroHeader
  className="tag_index_header tag_index_header--has_image"
  image={
    Object {
      "size": Object {
        "src": "testSrc",
      },
    }
  }
>
  <h1
    className="tag_index_header__title tag_index_header__title--has_image"
  >
    Test title
  </h1>
  <span
    className="tag_index_header__body tag_index_header__body--has_image"
  >
    50 posts in this category
  </span>
</HeroHeader>
`;
